### ISSUE TITLE
Document deprecation policy and how to implement a deprecation

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8082]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8082|
|Type|enhancement|
|Created|2015-11-04 10:06:57Z|
|Branch|https://github.com/twisted/twisted/tree/deprecation-docs-8082-2|

We have a wiki page which is WIP https://twistedmatrix.com/trac/wiki/CompatibilityPolicy

We should move it into the narrative docs and make it official.

<details><summary>Searchable metadata</summary>

```
trac-id__8082 8082
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__branches_deprecation_docs_8082_2 branches/deprecation-docs-8082-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__None None
time__1446631617347318 1446631617347318
changetime__1460883389418910 1460883389418910
version__None None
owner__adiroiban adiroiban
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46049]) Branching to deprecation-docs-8082.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban removed owner|
|-|-|

# Scope

This moved the WIP https://twistedmatrix.com/trac/wiki/CompatibilityPolicy wiki page to a final documentation which is part of Twisted documentation.

# Changes

Most of the contest is a direct copy of the wiki page.

I have re-arranged the compatible and non-compatible examples.

I added example for how to use @deprecated and deprecatedModuleAttribute together with direct calls to warnings.warn

I have extended the unit testing part with examples of how to use those methods

I have inserted a reference to trial bug related to not failing on un-handled deprecation errors and I tried to add a workaround.

The wiki talks about a 3 phase deprecation process but it only describes 2 phases. I have updated the text to say 2 phases.

I would say the current way of checking deprecation warnings using self.flushWarnings() is bad and we should consolidate all those calls into a single helper method... but I have just went for recommendation found in the wiki page. This can be changed later.


# How to test

Check that changes make sense... I have sent the branch to buildbot and there should be an HTML

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

I forgot that I have also created [#7894](https://github.com/twisted/twisted/issues/7894) for the same issue

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this.

Few notes:

- Line 451 on the new docs, the flushWarnings should really specify the test case.
- Line 419 down -- 1 sentence per line, please.
- Line 509 -- Typo 'trail' -> 'Trial'.
- The docstring probably wants to say 'now', not 'not'?
- The index should have it title cased ("Compatibility Policy")

Please fix these issues and put up for re-review. Thanks! <3

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban removed owner|
|-|-|

Thanks for the review.

Please check latest changes.


> Line 451 on the new docs, the flushWarnings should really specify the test case.

True. I have also updated the docstring.


> Line 419 down -- 1 sentence per line, please.


Done

> Line 509 -- Typo 'trail' -> 'Trial'.

Done.

> The docstring probably wants to say 'now', not 'not'?

At which line? At 510 is 'not'. As tests will not fail.

> The index should have it title cased ("Compatibility Policy")

Done.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this.

Rather than going through a few more cycles for minor changes, I went in and did them on the branch -- please check over my commits, and merge if you're happy, otherwise change them and put up for review.

Thanks! <3

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Many thanks for the changes. All good... but I found that the @deprecated decorator does not play nice with @property decorator.

I don't know how to fix it.

Please suggest a possible fix.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

[#8124](https://github.com/twisted/twisted/issues/8124) is the ticket following up on the @property fixes.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban removed owner|
|-|-|

I have updated the docs now that [#8124](https://github.com/twisted/twisted/issues/8124) is fixed.

How to deprecate instance attributes for old style classes ?

Please check that class deprecation is the right way to do it... not sure if deprecateModuleAttirbute is the right way and whether it should be call before the class definition, inside the class or after the class :)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Just spun the builders to check syntax &c

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for contributing this, Adi.  You're absolutely right that this should move into official policy documentation and stop rotting on the wiki.  When this lands, please delete the wiki page and replace it with a link to the document :).

Replying to [adiroiban](#note_10):
> I have updated the docs now that [#8124](https://github.com/twisted/twisted/issues/8124) is fixed.

Please merge forward so that the builders can succeed :-\.

> How to deprecate instance attributes for old style classes ?

Right now, there's no way.  I think it's fine not to talk about it; this document can land before being perfect.  File a new ticket for this functionality and let's discuss there.

> Please check that class deprecation is the right way to do it... not sure if deprecateModuleAttirbute is the right way and whether it should be call before the class definition, inside the class or after the class :)

This is correct documentation of the state of the art.  It should not be called inside the class, but either before or after should be OK.

However, it would be better if `@deprecated` did the work of setting up `deprecatedModuleAttribute` for you, since both importing and calling a deprecated name should raise warnings.  Separate ticket though! :)  Now that these docs are landed, we can have that ticket update the narrative documentation explaining how to use it as well, which is all the more reason to land this change!

> Please review.

This is looking very good.  Please merge forward and if you don't encounter any issues getting an all-green build, go ahead and land.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

(In [47264]) Branching to deprecation-docs-8082-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban commented|
|-|-|

Thanks for the review.

I have created [#8297](https://github.com/twisted/twisted/issues/8297) to follow up on deprecated instances for old style classes.

Looking forward for the buildbot results and will merge.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban set status to `closed`|
|-|-|

(In [47279]) Merge deprecation-docs-8082-2: Document the deprecation policy.

Author: adiroiban
Reviewer: hawkowl, glyph
Fixes: [#8082](https://github.com/twisted/twisted/issues/8082)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban commented|
|-|-|

I think that wiki page should be updated after the release... for now I left a comment at the top of that page
